plutus-ledger-api/test-onchain-evaluation/Main.hs

```diff
@@ -19,46 +19,53 @@ import System.Environment (getEnv)
 import System.FilePath (isExtensionOf)
 import Test.Tasty
 import Test.Tasty.HUnit
+import Test.Tasty.Providers
+
+data SingleTest
+    = SingleTest
+        LedgerPlutusVersion
+        (Maybe (EvaluationContext, [Integer]))
+        ScriptEvaluationEvent
+
+instance IsTest SingleTest where
+    run _opts (SingleTest ver mbCtx ev) _reportProgress = case mbCtx of
+        Just (ctx, params) ->
+            evaluate
+                =<< ( pure . maybe (testPassed mempty) (testFailed . display) $
+                        checkEvaluationEvent ctx params ev
+                    )
+        Nothing -> pure . testFailed $ "missing cost parameters for " <> show ver
+
+    testOptions = mempty
 
 -- | Test cases from a single event dump file
-testOneFile :: FilePath -> IO TestTree
-testOneFile eventFile = do
-    events <- CBOR.readFileDeserialise @ScriptEvaluationEvents eventFile
-    ctxV1 <-
-        either (assertFailure . display) pure $
-            mkContext V1.mkEvaluationContext (eventsCostParamsV1 events)
-    ctxV2 <-
-        either (assertFailure . display) pure $
-            mkContext V2.mkEvaluationContext (eventsCostParamsV2 events)
-    pure . testGroup eventFile $
-        fmap
-            ( \(idx, event) ->
-                testCase ("test case " <> show idx) $
-                    evaluate =<< runTest ctxV1 ctxV2 event
-            )
-            (zip [1 :: Int ..] (toList (eventsEvents events)))
+testOneFile :: TestName -> ScriptEvaluationEvents -> TestTree
+testOneFile name events =
+    case ( mkContext V1.mkEvaluationContext (eventsCostParamsV1 events)
+         , mkContext V2.mkEvaluationContext (eventsCostParamsV2 events)
+         ) of
+        (Right ctxV1, Right ctxV2) ->
+            testGroup name $
+                fmap
+                    ( \(idx, event) ->
+                        singleTest ("test case " <> show idx) $ case event of
+                            PlutusV1Event{} -> SingleTest PlutusV1 ctxV1 event
+                            PlutusV2Event{} -> SingleTest PlutusV2 ctxV2 event
+                    )
+                    (zip [1 :: Int ..] (toList (eventsEvents events)))
+        (Left err, _) ->
+            testCase name . assertFailure $ display err
```

Comment:
Interesting. This wouldn't have occurred to me! I actually don't know what the nicest way to handle failures that occur while building the tree of tests is. What happens if you leave it in IO as it was and just throw an exception?

Reply:
> What happens if you leave it in IO as it was and just throw an exception?

Then no test would be run, if there's a single event file for which we can't construct the EvaluationContext.

This, on the other hand, has the desired behavior: it will just fail that event file, it will say "this file failed", but other tests will still run.

Reply:
> I actually don't know what the nicest way to handle failures that occur while building the tree of tests is

Yeah, each single test can run IO, but it seems you can't run IO when combining a bunch of test into a test group, so I don't know. Ideally, you also don't want `CBOR.readFileDeserialise` failure on one file to affect other files.

Reply:
`defaultMain` takes `TestTree`, so you'd need to run the IO before `defaultMain` which is what I'm doing. And in a `TestTree`, you can run IO but only on the leaf test cases, not on test groups, which is what we really need. `withResource` doesn't seem to help either.

Reply:
Why not? It lets you run an IO action that (apparently) gets run by tasty at some point. So maybe tasty handles exceptions thrown by that action gracefully. I don't know, but it seemed worth looking.